mlem/contrib/kubernetes/base.py

```diff
@@ -0,0 +1,247 @@
+            host = ingress.hostname or ingress.ip
+        return HTTPClient(host=host, port=port)
+
+
+class K8sEnv(MlemEnv[K8sDeployment]):
+    type: ClassVar = "kubernetes"
+    deploy_type: ClassVar = K8sDeployment
+    registry: Optional[DockerRegistry] = None
+
+    def create_k8s_resources(self, generator):
+        k8s_client = client.ApiClient()
+        with tempfile.TemporaryDirectory(
+            prefix="mlem_k8s_yaml_build_"
+        ) as tempdir:
+            filename = os.path.join(tempdir, "resource.yaml")
+            generator.write(filename)
+            utils.create_from_yaml(k8s_client, filename, verbose=True)
+
+    def deploy(self, meta: K8sDeployment):
+        self.check_type(meta)
+        redeploy = False
+        with meta.lock_state():
+            meta.load_kube_config()
+            state: K8sDeploymentState = meta.get_state()
+            if state.image is None or meta.model_changed():
+                image_name = (
+                    meta.image_name or generate_docker_container_name()
+                )
+                echo(EMOJI_BUILD + f"Creating docker image {image_name}")
+                if self.registry is None and meta.registry is None:
+                    echo(
+                        EMOJI_FAIL
+                        + "Error: No registry set to be used by Docker"
+                    )
+                    raise ValueError(
+                        "registry to be used by Docker is not set or supplied"
+                    )
+                registry = (
+                    meta.registry
+                    if meta.registry is not None
+                    else self.registry
+                )
+                with set_offset(2):
+                    state.image = build_model_image(
+                        meta.get_model(),
+                        image_name,
+                        meta.server
+                        or project_config(
+                            meta.loc.project if meta.is_saved else None
+                        ).server,
+                        DockerEnv(registry=registry, daemon=meta.daemon),
+                        force_overwrite=True,
+                        # runners usually do not support arm64 images built on Mac M1 devices
+                        platform="linux/amd64",
+                    )
+                meta.update_model_hash(state=state)
+                redeploy = True
+
+            if state.deployment_name is None or redeploy:
+                k8s_yaml_builder = K8sYamlBuilder(
+                    image=state.image,
+                    image_pull_policy=meta.image_pull_policy,
+                    port=meta.port,
+                    service_type=meta.service_type,
+                )
+                generator = K8sYamlGenerator(
+                    **k8s_yaml_builder.get_resource_args().dict()
+                )
+                self.create_k8s_resources(generator)
+
+                w = watch.Watch()
+                for event in w.stream(
```

Comment:
also a good candidate for separate method. and also make timeout configurable

Reply:
done